include/seahorn/Expr/EvalModel.hh

```diff
@@ -0,0 +1,302 @@
+#pragma once
+
+#include "seahorn/Expr/EvalUtils.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprGmp.hh"
+#include "seahorn/Expr/ExprOpBind.hh"
+#include "seahorn/Expr/ExprOpBinder.hh"
+
+#include <math.h>
+#include <stdlib.h>
+
+namespace expr {
+namespace eval {
+
+/**
+ * Bit vector structure. Holds the number(as type T) and its width
+ *
+ * \note bools are stored as 1/0 with a width of 1
+ */
+template <typename T> struct BvNum {
+  T num = 0;
+  unsigned width = 0;
+
+  /// \brief empty constructor
+  BvNum() : num(0), width(0) {}
+
+  /// \brief bool constructor.
+  BvNum(bool a) : num(a), width(1) {}
+
+  BvNum(T numArg, unsigned widthArg) : num(numArg), width(widthArg) {}
+
+  bool operator<(const BvNum<T> &bnum) const { return num < bnum.num; }
+
+  bool operator==(const BvNum<T> &bnum) const {
+    return (num == bnum.num) && (width == bnum.width);
+  }
+};
+
+/**
+ * Holds the data for an array.
+ *
+ * In EvalModel, the array expressions are mapped to this struct
+ *
+ */
+template <typename T> struct Arr {
+  bool m_hasDefaultVal = false;
+  BvNum<T> m_defaultVal;
+
+  unsigned m_valWidth = 0;
```

Comment:
what is `m_valWidth`? isn't width already in `BvNum`?

Reply:
The purpose of m_valWidth is for generating new BvNums. If you try to generate a new number when there isn't any BvNums stored yet then you won't know how wide to make it

Reply:
Maybe I could just store the width in m_valWidth and store the values only as the number and not bvnum to reduce some overhead?